Parser_BacklinkCount modified

jedelbo opened the pull request.

ironage commented:
Here's a visualisation, with the link numbers referring to the `item_id` and the backlinks referring either to the `item_id` or `customer_id` depending on where they link from:
`class_Items`

item_id  | self | self_backlinks | items_backlinks | fav_item_backlinks | @links.@count
--------|-----|---------------|-----------------|-------------------|----------------
0	|	0 |	0	|		0,2,2,2,2,2,2,2,2,2,2	|	customer_id=4	| 13
1	|	1 |	1	|		0					| customer_id=2	| 3
2	|	2 |	2	|		0,4,4				|	customer_id=0	| 5
3	|	3 |	3	|							|	|	1 
4	|	   |		|							|	|	0

`class_Person`

customer_id |   items   			| fav_item
-------------|--------------------|---------
0		|	0,1,2			| 2
2		|	0,0,0,0,0,0,0,0,0,0	| 1
4		|	2,2				| 0


Unless I'm misunderstanding you, I think the functionality is correct. I think what's throwing the intuition off is that `@links.@count`is a special syntax which always refers to the total backlink count of that row (the sum of entries in all backlink columns). The changed check `fav_item.@links.@count` goes through a forward link but it does not refer to the backlinks of only that forward link, it still computes the backlink totals just on the table referred to by the link. To ask for the backlinks of a particular column we would use `fav_item.@links.Person.@count`. Testing this other syntax is located in another test. Does that clarify?

jedelbo commented:
I have changed the test a bit again to match the test I ended up having in realm-core-private. The problem is that you do not follow the 'fav_item' link. The original test did not expose this problem as the linked object has the same index in the table as the linking object. When you cross link, you will se the problem.
I have added a commit that solves the problem.

ironage commented on test/test_parser.cpp:
```diff
@@ -1840,35 +1842,35 @@ TEST(Parser_BacklinkCount)
-    verify_query(test_context, items, "@links.class_Person.fav_item.items.@links.@count == 1 && item_id == 3", 1);
-    verify_query(test_context, items, "@links.class_Person.fav_item.fav_item.@links.@count == 1 && item_id == 3", 1);
+    verify_query(test_context, items, "@links.class_Person.fav_item.items.@links.@count == 130 && item_id == 2", 1);
+    verify_query(test_context, items, "@links.class_Person.fav_item.fav_item.@links.@count == 3 && item_id == 2", 1);
```
These two checks are not correct actually, as far as I understand we shouldn't sum the backlinks at each stage of the link, just return the backlink count at the destination table. I think it was correct before?

ironage commented on test/test_parser.cpp:
```diff
@@ -1880,8 +1882,7 @@ TEST(Parser_BacklinkCount)
 
     // backlink count compared to link count
     verify_query(test_context, items, "@links.@count >= self.@count", 5);
-    verify_query(test_context, t, "items.@count >= items.@links.@count", 1);
```
I think it could still be valuable to test over a single link and over a list, and they should functionally give the same result, right?
